agimus_controller/agimus_controller/mpc.py

```diff
@@ -0,0 +1,63 @@
+import time
+from agimus_controller.ocp_base import OCPBase
+from agimus_controller.warm_start_base import WarmStartBase
+from agimus_controller.trajectory import (
+    TrajectoryPoint,
+    WeightedTrajectoryPoint,
+    TrajectoryBuffer,
+)
+from agimus_controller.mpc_data import MPCResults, MPCDebugData
+
+
+class MPC(object):
+    def __init__(self) -> None:
+        self._ocp = None
+        self._warm_start = None
+        self._mpc_results: MPCResults = None
+        self._mpc_debug_data: MPCDebugData = None
+        self._trajectory_buffer = TrajectoryBuffer()
+
+    def setup(self, ocp: OCPBase, warm_start: WarmStartBase) -> None:
+        self._ocp = ocp
+        self._warm_start = warm_start
+
+    def run(self, initial_state: TrajectoryPoint, current_time_ns: int) -> MPCResults:
+        assert self._ocp is not None
+        assert self._warm_start is not None
+        timer1 = time.perf_counter_ns()
+        self._clear_buffer_past(current_time_ns)
+        reference_trajectory = self._extract_horizon_from_buffer()
+        self._ocp.set_reference_horizon(reference_trajectory)
```

Comment:
I think that we shouldn't reset the ocp by giving every time the whole horizon of references / weights, for two reasons : 
- the references for each node simply are the references of the next node and we just have to input the references of the last node
- for weights, I think we shouldn't let people change all the weights from an iteration to another. Because if we do that, the mpc will lose his ability to predict how to move, since the problem change at every iteration

Reply:
Is the current API limiting you in doing what you proposed? The proposed interface should be general enough to not limit you in what you do. If I send a completely different trajectory, then OCP will not be optimal, but it will compute output, and I can use it - I just have to expect worse performance or even failure if it does not converge. The same is true for weight; if you want them to be fixed, then just send constant weights - API is not limiting you in doing so.

Reply:
the current API isn't limiting me in what I proposed, I understand that you want something very general, but in this case, I'm afraid we only open doors to mpc controls that wouldn't work well, so I'm not sure in this case that it's something we want

Reply:
I changed my mind and I think having this API could be useful https://github.com/agimus-project/agimus_controller/pull/97#discussion_r1873499242

Reply:
Ok. But that case is implementation-specific, and that is something you take care of from outside this code. So the API is generic enough for everyone's implementation, while you can tailor it to your cause the you prefer.

If someone wants to do something stupid, then fine, they break the robot. This is designed to be used by someone who knows what is happening and what might be the consequences of changing weights.

In your first bullet point, you mean that the underlying implementation in Crocoddyl uses shared pointer and all we need it to roll the references, by reattaching references and then adding new one at the end, meaning it's pure performance concern? If so, then I agree it will be more efficient, but I do not believe this will slow us down that much, while this will heavily constrain us on more advanced handling of the buffer. From an implementation point of view, it's nice to have, but I don't believe it will hit the performance a lot

Reply:
So the question is close?
@TheoMF if you are ok with these explanation can you click on "resolve conversation"?

Reply:
> In your first bullet point, you mean that the underlying implementation in Crocoddyl uses shared pointer and all we need it to roll the references, by reattaching references and then adding new one at the end, meaning it's pure performance concern?

There is the performance side but also I think starting to change all the references of the nodes will most of the time end up in bad control, but there's cases where it could be useful I guess

Reply:
Usually this will be simply changing references to "new" ones with the exact same values so in practice we will end up with exact same problem from maths point of view